staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go

```diff
@@ -448,6 +448,12 @@ func (c *Cacher) startCaching(stopChannel <-chan struct{}) {
 	c.watchCache.SetOnReplace(func() {
 		successfulList = true
 		c.ready.set(true)
+		// We rely on the fact that onReplace callback is called
+		// with watchCache lock being held.
+		// That ensures that Watch can't be registered (as it also
+		// grabs the watchcache lock) and ensures no deadlock as
```

Comment:
I'm fuzzy about the deadlock being prevented... was that a deadlock we were vulnerable to before?

In addition to the comments explaining why it's safe to call terminateAllWatchers() here, maybe explain why we need to terminate watchers again?